### ISSUE TITLE
Segmentation fault on 32-bit ubuntu 10.04

### ISSUE THREAD

**Issue by unrealsolver** [wkhtmltopdf/wkhtmltopdf#2160]
32-bit build crashes (on any page), however 64-bit one works fine.
I'm using Jan 9, 2015  source release. It also fails on previous releases.
```
vagrant@vagrant:~$ wkhtmltopdf test.html  out.pdf
Loading pages (1/6)
Segmentation fault=============>                             ] 50%
vagrant@vagrant:~$ wkhtmltopdf -V
wkhtmltopdf 0.12.2 (with patched qt)
vagrant@vagrant:~$ uname -a
Linux vagrant 2.6.32-38-generic #83-Ubuntu SMP Wed Jan 4 11:13:04 UTC 2012 i686 GNU/Linux
```

**Comment by ashkulz** [comment]
Did you compile it from scratch using the `posix-local` target?

**Comment by unrealsolver** [comment]
Yes, I built it from scratch using `posix-local`.

**Comment by unrealsolver** [comment]
More info
tl;dr: Looks like problem with XListFonts () from /usr/lib/libX11.so.6
```
(gdb) run
Starting program: /usr/local/bin/wkhtmltopdf test.html out.pdf
[Thread debugging using libthread_db enabled]
[New Thread 0xb7e82b70 (LWP 25872)]
[New Thread 0xb7559b70 (LWP 25873)]
Loading pages (1/6)
[==============================>                             ] 50%
Program received signal SIGSEGV, Segmentation fault.
0x0021fb97 in XListFonts () from /usr/lib/libX11.so.6
(gdb) run
The program being debugged has been started already.
Start it from the beginning? (y or n) y
Starting program: /usr/local/bin/wkhtmltopdf test.html out.pdf
[Thread debugging using libthread_db enabled]
[New Thread 0xb7e82b70 (LWP 25875)]
[New Thread 0xb7559b70 (LWP 25876)]
Loading pages (1/6)
[==============================>                             ] 50%
Program received signal SIGSEGV, Segmentation fault.
0x0021fb97 in XListFonts () from /usr/lib/libX11.so.6
(gdb) backtrace
#0  0x0021fb97 in XListFonts () from /usr/lib/libX11.so.6
#1  0x0953b209 in loadXlfds(char const*, int) ()
#2  0x0953bb65 in load(QString const&, int, bool) ()
#3  0x0953f3c0 in QFontDatabase::families(QFontDatabase::WritingSystem) const ()
#4  0x0953f71e in QFontDatabase::hasFamily(QString const&) const ()
#5  0x086c722c in WebCore::FontCache::createFontPlatformData(WebCore::FontDescription const&, WTF::AtomicString const&) ()
#6  0x085008f8 in WebCore::FontCache::getCachedFontPlatformData(WebCore::FontDescription const&, WTF::AtomicString const&, bool) ()
#7  0x08500fb3 in WebCore::FontCache::getFontData(WebCore::Font const&, int&, WebCore::FontSelector*) ()
#8  0x084f4482 in WebCore::FontFallbackList::determinePitch(WebCore::Font const*) const ()
#9  0x085bd34d in WebCore::RenderBlock::LineBreaker::nextLineBreak(WebCore::BidiResolver<WebCore::InlineIterator, WebCore::BidiRun>&, WebCore::LineInfo&, std::pair<WebCore::RenderText*, WebCore::LazyLineBreakIterator>&, WebCore::RenderBlock::FloatingObject*) ()
#10 0x085bff02 in WebCore::RenderBlock::layoutRunsAndFloats(bool, bool, WTF::Vector<WebCore::RenderBlock::FloatWithRect, 0u>&, int&, int&) ()
#11 0x085c1bd5 in WebCore::RenderBlock::layoutInlineChildren(bool, int&, int&) ()
#12 0x085b0017 in WebCore::RenderBlock::layoutBlock(bool, int) ()
#13 0x08593e20 in WebCore::RenderBlock::layout() ()
#14 0x085aee5d in WebCore::RenderBlock::layoutBlockChild(WebCore::RenderBox*, WebCore::RenderBlock::MarginInfo&, int&, int&) ()
#15 0x085af5af in WebCore::RenderBlock::layoutBlockChildren(bool, int&) ()
#16 0x085b04ea in WebCore::RenderBlock::layoutBlock(bool, int) ()
#17 0x08593e20 in WebCore::RenderBlock::layout() ()
#18 0x085aee5d in WebCore::RenderBlock::layoutBlockChild(WebCore::RenderBox*, WebCore::RenderBlock::MarginInfo&, int&, int&) ()
#19 0x085af5af in WebCore::RenderBlock::layoutBlockChildren(bool, int&) ()
#20 0x085b04ea in WebCore::RenderBlock::layoutBlock(bool, int) ()
#21 0x08593e20 in WebCore::RenderBlock::layout() ()
#22 0x08679b58 in WebCore::RenderView::layout() ()
#23 0x084c86d3 in WebCore::FrameView::layout(bool) ()
#24 0x0824a068 in WebCore::Document::implicitClose() ()
#25 0x084396d4 in WebCore::FrameLoader::checkCallImplicitClose() ()
#26 0x0844454f in WebCore::FrameLoader::checkCompleted() ()
#27 0x0844470b in WebCore::FrameLoader::finishedParsing() ()
#28 0x08241b77 in WebCore::Document::finishedParsing() ()
#29 0x083ca0e6 in WebCore::HTMLTreeBuilder::finished() ()
#30 0x083b101d in WebCore::HTMLDocumentParser::prepareToStopParsing() ()
#31 0x083b04bd in WebCore::HTMLDocumentParser::finish() ()
#32 0x082323cb in WebCore::Document::finishParsing() ()
#33 0x08433e73 in WebCore::DocumentWriter::endIfNotLoadingMainResource() ()
#34 0x0842c789 in WebCore::DocumentLoader::finishedLoading() ()
#35 0x0844507a in WebCore::FrameLoader::finishedLoading() ()
#36 0x0845892e in WebCore::MainResourceLoader::didFinishLoading(double) ()
#37 0x08469d98 in WebCore::ResourceLoader::didFinishLoading(WebCore::ResourceHandle*, double) ()
#38 0x086b84e5 in WebCore::QNetworkReplyHandler::finish() ()
#39 0x086b6793 in WebCore::QNetworkReplyHandlerCallQueue::flush() ()
#40 0x086b6c9a in WebCore::QNetworkReplyWrapper::didReceiveFinished() ()
#41 0x086b9c20 in WebCore::QNetworkReplyWrapper::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) ()
---Type <return> to continue, or q <return> to quit---
```

**Comment by unrealsolver** [comment]
Recently I've noticed same issue with 64-bit ubuntu. It works locally with same version of libX11 (6.3.0). Also I've checked size and md5hash - no differences.
That means problem is with my build script or environment (wkhtmltopdf binaries have different sizes on different hosts (both 64 bit)).

**Comment by ashkulz** [comment]
Considering that Ubuntu 10.04 is EOL in a few days, I don't see a point in investigating this. Please reopen if it is reproducible on 12.04 or later.

**Comment by fdelapena** [comment]
Same issue with Fedora 22 using latest git.

**Comment by ashkulz** [comment]
I think someone faced this with a custom build on Amazon Linux recently, workaround was to run it under xvfb.

**Comment by fdelapena** [comment]
Thanks, I was trying to run it from a remote ssh session, then looks like headless support is somewhat broken here, too.

Passing --use-xserver worked properly when using ssh -X. Also, prefixing xvfb-run worked when running in remote ssh without -X.

**Comment by ashkulz** [comment]
@fdelapena: can you try if [this download](https://bitbucket.org/wkhtmltopdf/wkhtmltopdf/downloads/wkhtmltox-0.12.3-dev-4c066d2.tar.xz) works for you? I'm trying out a generic linux build based on CentOS 6, so maybe that will work for you...

**Comment by fdelapena** [comment]
Thank you @ashkulz, the provided binary works properly. Tested on remote Fedora 22 via SSH without xvfb and without X11 forwarding and it works without dependency issues and the exported PDF looks properly:
```
$ ./wkhtmltopdf http://www.google.com/ googletest.pdf
Loading pages (1/6)
Counting pages (2/6)                                               
Resolving links (4/6)                                                       
Loading headers and footers (5/6)                                           
Printing pages (6/6)
Done
```

```
$ ldd wkhtmltopdf 
	linux-vdso.so.1 (0x00007ffd49fe2000)
	libXrender.so.1 => /lib64/libXrender.so.1 (0x00007f96a970a000)
	libfontconfig.so.1 => /lib64/libfontconfig.so.1 (0x00007f96a94c7000)
	libfreetype.so.6 => /lib64/libfreetype.so.6 (0x00007f96a921c000)
	libXext.so.6 => /lib64/libXext.so.6 (0x00007f96a900a000)
	libX11.so.6 => /lib64/libX11.so.6 (0x00007f96a8cc9000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f96a8ab3000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f96a88af000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f96a86a7000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f96a848b000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f96a8109000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f96a7e01000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f96a7bea000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f96a782a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f96a9914000)
	libexpat.so.1 => /lib64/libexpat.so.1 (0x00007f96a7600000)
	libbz2.so.1 => /lib64/libbz2.so.1 (0x00007f96a73f0000)
	libpng16.so.16 => /lib64/libpng16.so.16 (0x00007f96a71bd000)
	libxcb.so.1 => /lib64/libxcb.so.1 (0x00007f96a6f9b000)
	libXau.so.6 => /lib64/libXau.so.6 (0x00007f96a6d97000)
```

I tried to run the existing prebuilt stable binary download (0.12.2.1 for CentOS 7.0.1406) on the website but the libpng was trying to find libpng15, which was not working here. The one you provided works without issues :+1:.

**Comment by ashkulz** [comment]
Yeah, it statically links in `libpng` and `libjpeg` as they have different SONAMEs on different distributions. BTW `ldd` is not a perfect way to find out dependencies, `objdump -p <binary> | grep NEEDED` is the preferred way -- `ldd` uses knowledge of installed libraries to report the dependencies which is not always correct (e.g. `libpng16` is actually required by `libfreetype6` and not by wkhtmltopdf, which uses it statically linked).

The downside is that you **have** to update wkhtmltopdf whenever a security issue is discovered in the statically linked libraries i.e. `libpng` and `libjpeg`. Such "generic" linux binaries should be available from the official downloads page from the 0.12.3 release onwards.

**Comment by ashkulz** [comment]
A preview for the next release [`0.12.3-dev-79ff51e`](http://wkhtmltopdf.org/downloads.html#testing) is available, which has generic linux binaries with minimal dependencies. Please report back whether it works for you without any issues.

*Please note that the above downloads will be removed after the `0.12.3` release*.

**Comment by fdelapena** [comment]
Thanks @ashkulz, tested on remote headless Debian 8 and Fedora 22, both worked properly :+1:.
